### DISCUSSION TITLE
Playing audio (mp3) via PWM

### DISCUSSION THREAD

**Discussion by tank104** [nanoframework/Home#1190] [General]
Hi,
Sorry if the wrong place to ask, but looking to play audio files via PWM on my dev board.
I can see PWM here:
https://github.com/nanoframework/System.Device.Pwm

but can find nothing on loading mp3/wav/ogg file and certainly nothing on playing it via pwm.

Trying to achieve something like this:
https://learn.adafruit.com/mp3-playback-rp2040

Any help at all appreciated!

thanks,

andrew

**Comment by Ellerbach** [comment]
There is no class to load mp3 or ogg. If you're using PWM, it's recommended you use tones. See the buzzer as an example. It can play melodies: https://github.com/nanoframework/nanoFramework.IoT.Device/tree/develop/devices/Buzzer
Wav is simple and well known. Checkout the I2S sample pack to on how to open them: https://github.com/nanoframework/Samples/tree/main/samples/I2S

Note that in all cases, the quality you'll get with PWM won't be good compared to using I2S (if your device supports it, so far only implements on ESP32).

**Reply by Ellerbach** [reply]
And note that the play back on the RP2040 you pointed out is I2S, not PWM.

**Reply by tank104** [reply]
Thanks so much for that I2S link - I looked thoroughly through the samples page but could find nothing with audio/mp3/wav - but didn't think to look for I2S.

The reason I though that Python example was PWM is because of the object "PWMAudioOut".  I see further down the page it mentioned I2S - so will look into that too.

Really appreciate your help! Just ordered a batch of ESP32-C3-DevKitC-02 boards so sounds like should be good.

Keen to get away from Python and into c# that I am far more comfortable with.

**Reply by tank104** [reply]
The more I look into it the more I am sure that example is using PWM.
They have a seperate example for I2s here: https://learn.adafruit.com/mp3-playback-rp2040/pico-i2s-mp3

Realise you have been helpful enough at the moment, but just wanted to check there isn't a way to play wav files via PWM?

I can see how to do tones, but not put the pieces together to stream the WAV to PWM

**Reply by Ellerbach** [reply]
We don't have a direct way to play the WAV into PWM. I doubt about the end quality anyway. But I do understand what you want.
You would have I think to have a period 2 times higher than the once in the file because then, you want to convert a 0-255 value into a percentage of a period.
Let's say your file if 8 MHz, 8 bit. Then the PWM would have to be 16 MHz and you would have to change the signal all the time. And at this frequency, I'm not sure of the quality but you can give it a try!

**Reply by tank104** [reply]
Thanks!
I have found some i2s amps, so will just swap out for the PWM amps I had brought.

appreciate your time
